### ISSUE TITLE
SSDP blasts and listen are on 'default' interface, needs to be smart or configurable

### ISSUE THREAD

**Issue by mooninite** [jellyfin/jellyfin#2986]
**Describe the bug**
I run Jellyfin on a server that also acts as an Internet router.
- Interface eth0 - the Internet
- Interface eth1 - the LAN

Jellyfin network settings: Secure mode is set to Disabled. Local address is set to the LAN IP or blank.

The DLNA SSDP blasts and listener are sent over eth0 no matter what Jellyfin network setting. This causes the Jellyfin server to not appear as a discoverable server on my LAN.

**System (please complete the following information):**
 - OS: Fedora
 - Virtualization: None
 - Clients: Firefox, Android, DLNA clients
 - Browser: Firefox 75
 - Jellyfin Version: 10.5.4
 - Playback: n/a
 - Installed Plugins: n/a
 - Reverse Proxy: Apache
 - Base URL: None
 - Networking: Direct - no VM or container
 - Storage: local

**To Reproduce**

1. Start Jellyfin
2. sudo tcpdump -ni eth0 udp and port 1900 -A
3. See SSDP messages from Jellyfin
4. sudo tcpdump -ni eth1 udp and port 1900 -A
5. Never see SSDP messages from Jellyfin

**Expected behavior**

Jellyfin should have an option to define the network interface it broadcasts on or else use the provided LAN IP to identify the interface name.

**Logs**

**Screenshots**

**Additional context**

**Comment by mooninite** [comment]
After digging around I discovered there is somewhat of a configurable switch, but not an ideal one.

PR #978 added two config-file only switches. The /etc/jellyfin/dlna.xml / SendOnlyMatchedHost option allows Jellyfin to send out messages on all interfaces. While this "fixes" it the multicast messages are still being sent out over the Internet interface.

I see that the issue to create a web UI option for these new option was abandoned, but it should probably be brought back up in addition to adding an option or automatic logic to send out over the correct network interface.

**Comment by BaronGreenback** [comment]
what are your interface types and what are the ip addresses - so i can recreate here?

**Comment by mooninite** [comment]
eth0 - Ethernet - Intel i210 NIC - Public IP
eth1 - Ethernet - Intel i210 NIC - 10.0.0.0/24 network IP

**Comment by BaronGreenback** [comment]
both real adapters? Also, have you got any settings for LAN networks, or bind in jellyfin options?

**Comment by mooninite** [comment]
Yes, both are real, physical adapters. No VMs, containers, docker, or any other virtual tech is in use. 

LAN networks: 127.0.0.1, 10.0.0/24
Bind to local network address: <blank> - I have also tried setting my LAN IP here, makes no difference.
Default ports
Secure connection mode: Disabled

**Comment by BaronGreenback** [comment]
Looks like the problem is with the virtual interface detection mechanism in the networkManager code - function GetIPsDefault.

Because the second interface doesn't have a gateway, the function ignores the lan address.

**Comment by BaronGreenback** [comment]
I see a few issues in the code:-

1) The blast being send of eth0 not eth1 is due to eth1 not having a next hop address. (GetLocalIPSubnetMask and GetIPsDefault)
2) eth0 is being used, because the LAN subnets are never checked to see if that interface should be excluded.

Am putting together a submission.

Can you think of any occasion when a Blast would need to be sent down a non-LAN address?

**Comment by mooninite** [comment]
Thanks for looking into this. Short answer: No, we should not send a blast on non-LAN addresses. Long answer: A blast is only good for local-link type clients to receive so sending one out a non-LAN address will never make it past the next hop anyway. There is one use-case I have seen but I do not feel Jellyfin should support it.

**Comment by BaronGreenback** [comment]
Two places in the code (one of which #978 modified) which i don't quite understand the reasoning behind. Have checked the history and it's been there for at least 3 years.

They use the GetIPProperties().GatewayAddresses.FirstOrDefault() and ignore the interface if it is null. 

https://github.com/jellyfin/jellyfin/blob/master/Emby.Server.Implementations/Networking/NetworkManager.cs/#L477 - lines 385 and 497

I've commented it out and am testing my mods to see if anything breaks.

The only things i can see that it seemed to have filtered out are loopback addresses, but if that were the case, there are much easier ways of doing so.

**Comment by BaronGreenback** [comment]
scratch that - it's for detecting VM interfaces

**Comment by BaronGreenback** [comment]
Have submitted the following change.

1. Minor: IsInPrivateAddressSpace - improved subnet code checking

2. Minor: Fixed an issue around invalid LAN address setting values causing the SSDP to crash (if you entered any text with a / in it SSDP process would bomb out)

3. Minor: corrected local link filter (FilterIPAddress) to filter on 169.254.X.X addresses instead of 169.X.
X.X

4. Major: filtered SSDP blasts from non-LAN addresses as defined on the network page.

5. Major: Fixed interfaces with no gateway being excluded from SSDP blasts (Lots of changes here)
removed #986 mod - as this was part of the issue of #2986. (This code was an attempt to exclude virtual interface adapters from SSDP and multicasts.)

As there is no sure way of performing this task automatically, i've switched it to manual.

If you enter an IP address within the LAN network settings in brackets it will be excluded from all network events (listeners, SSDP, and multicasts)

eg. 127.0.0.1, 10.0.0.0/24, [10.1.1.1]

will exclude the interface who's ip address is 10.1.1.1 from being "spoken" to.

**Comment by BaronGreenback** [comment]
Have submitted the following change.

1. Minor: IsInPrivateAddressSpace - improved subnet code checking

2. Major: Fixed interfaces with no gateway being excluded from SSDP blasts (Lots of changes here)
removed #986 mod - as this was part of the issue of #2986. (This code was an attempt to exclude virtual interface adapters from SSDP and multicasts.) 

The config property "ignoreVirtualInterface" now does nothing.

As there is no sure way of performing this task automatically, i've implimented the possility of doing this manually, by entering the IP address of the interface in brackets in the LAN settings config option. This will cause it to be excluded from all network events (listeners, SSDP) 

eg. 127.0.0.1, 10.0.0.0/24, [10.1.1.1] will exclude the traffic to the interface who's ip address is 10.1.1.1.

3. Major: filtered SSDP blasts from non-LAN addresses as defined on the network page.

4. Minor: Fixed an issue around invalid LAN address setting values causing the SSDP to crash (if you entered any text with a / in it SSDP process would bomb out)

5. Minor: corrected local link filter (FilterIPAddress) to filter on 169.254.X.X addresses instead of 169.X.X.X

Have tested on windows 7 64bit.

**Comment by mooninite** [comment]
Sorry for the delay in testing. I backported the PR for 10.5.5 and it is working much better for me. I no longer have to have any bind IP in my network settings and the broadcasts are going out over the LAN interface. The LAN IP is also being used for the DLNA Location / Base URL, which is good. However, I still see some broadcasts going out over the WAN interface. Not a deal breaker at this point though.

**Comment by BaronGreenback** [comment]
Just out of interest - try putting the external interface IP in the LAN list enclosed in [ ] brackets. eg [244.192.x.x] and see if stops broadcasts on the WAN. This tells the system not to use that interface for anything (it's was a manual way to filter out virtual interfaces) - will be interesting to see if it helps in your case.

If you could let me know the first two octets of the ip address I can try to replicate your setup here to investigate.

Aware of security concerns, but is that any traffic packet info on the broadcasts that you can share?

**Comment by mooninite** [comment]
Using brackets didn't stop it. I haven't taken a good look at the code, but I'm seeing Jellyfin bind to any IP for the DLNA port. Even with a LAN IP set for a bind IP it binds to any IP. My WAN IP is below, truncated, and UUID is redacted.

SSDP message:
```
14:06:44.012079 IP 47.24.x.x.59828 > 239.255.255.250.ssdp: UDP, length 352
E..|......../.0........l.hQ#NOTIFY * HTTP/1.1
HOST: 239.255.255.250:1900
DATE: Wed, 13 May 2020 19:06:43 GMT
CACHE-CONTROL: max-age = 1800
LOCATION: http://10.0.x.x:8096/dlna/<$UUID>/description.xml
SERVER: Linux/Unix 5.6.11.300 UPnP/1.0 RSSDP/1.0
NTS: ssdp:alive
NT: upnp:rootdevice
USN: uuid:<$UUID>::upnp:rootdevice
```
SSDP port:
```
$ sudo ss -anup | grep 59828
UNCONN 0      0                               0.0.0.0:59828       0.0.0.0:*      users:(("jellyfin",pid=1058378,fd=281))
```

**Comment by BaronGreenback** [comment]
Is sendOnlyMatchedHost set to true (the default) in your dnla config?
Does your internet facing router have SSDP enabled on its internal facing interface?

**Comment by mooninite** [comment]
> Is sendOnlyMatchedHost set to true (the default) in your dnla config?
It is set to false.
> Does your internet facing router have SSDP enabled on its internal facing interface?
No, it is a cable modem. It's a pure modem. No wi-fi and nothing "smart" about it.

**Comment by BaronGreenback** [comment]
I've been through the code and found a couple of other occations that SSDP traffic may leak out of the interface, due to a default listener being created for any ip address.

If an SSDP packet is received from a non-LAN port a response will be sent, and discovery messages will periodically go out regardless.

I've made modifications here - and will do some further testing tomorrow to ensure I haven't broken anything.

**Comment by BaronGreenback** [comment]
fixed in #3007

**Comment by mooninite** [comment]
Fixed in 10.6. Thanks.
